### DISCUSSION TITLE
workspaces: Enforce build order based on declared depedancies ?

### DISCUSSION THREAD

**Discussion by cassepipe** [pnpm/pnpm#6855] [Ideas]
Say let's say I have a fullstack monorepo with a `front`, `back` and `shared` package.

I need `shared` to build before  `front` and `back`, so I need to enforce some kind of build order.

Since both `front` and `back` declare `shared` as a dependency with the `workspace` protocol, couldn't there be a flag that would run a script respective of the dependency tree so that for example a `pnpm run build --ordered` from the workspace root would respect that build order ?

Coming from the C world I am surprised there's no way to ensure a basic build order int the npm world

**Comment by gluxon** [comment]
@cassepipe I think `pnpm recursive run --sort build` may help?

From `pnpm recursive run --help`:

```
Options:
...
      --sort                            Sort packages topologically (dependencies before dependents). Pass --no-sort to disable.
```

**Reply by cassepipe** [reply]
Haaa I  knew it had to exist. I missed that option, thank you so much !

**Reply by cassepipe** [reply]
Ha but it's a bit more complicated :
It's the default !
But it did not work for my project...
I need to investigate this

**Reply by gluxon** [reply]
Happy to help! Best of luck investigating.

**Comment by elitemike** [comment]
It's a shame this feature doesn't seem to work reliably.  I have a basic test project with 3 packages, A depends on B, B depends on C.  B is consistently being built before C.  

I'm evaluating my tools for a major rewrite of an application I build.  I'm currently using Yarn workspaces with wsrun to handle the order of building.  I was really hoping to have a few less addons in my pipeline.
